graph/kv/quad_iterator.go

```diff
@@ -123,6 +123,9 @@ func (it *quadIteratorNext) TagResults(dst map[string]graph.Ref) {}
 
 func (it *quadIteratorNext) Close() error {
 	if it.it != nil {
+		if err := it.it.Err(); err != nil && it.err == nil {
```

Comment:
This is not the right solution, I'm afraid. You'll need to find the "bad" iterator that doesn't propagate its error in `Close`. The code below already checks for that specifically.

And actually, the "bad" iterator is doing something worse: it's not recording `it.err` on `Next/Contains/NextPath`.  Otherwise usual iterator loop would catch it from `it.Err()`, instead of waiting for `Close` to return it. Usually we do not expect `Close` error to be handled, since it's usually too late.

If you would ask me to guess, I'll say the cause is probably somewhere in one of the `Contains` implementations.

Reply:
That may be so, but it's also valid to check Err the way it's done here, and foolproof, yes?

Reply:
Valid, but not foolproof, since it's just one iterator that does it. For it to be foolproof you'll need to go over ALL iterators and do the same change. It might be easier to locate a "bad" iterator and/or write a test or static check that does.

Reply:
I'll continue searching for the other bad iterators.

That said; it seems logical to check Err() here and set it if not nil, especially since the following two Close calls will overwrite it.err if set.

This is a good fix for the bug in the short term imo - thoughts?

Reply:
That's the thing: why `it.it.Close()` below doesn't return an error in the first place? This is the bug worth fixing.

If you have a case that you can debug, just do this (temporarily):

```go
err2 := it.it.Err()
if err := it.it.Close(); err != nil && it.err == nil {
	it.err = err
} else if err == nil && err2 != nil {
	panic(fmt.Errorf("incorrect implementation of %T.Close: Err must return same error as Close", it.it))
}
```

Once you trigger it, you know which `Close` needs to be fixed.